File: 2021_summer_algorithm_study_intermediate/using_cpp/2252.cpp
/*
input :
3 2
1 3
2 3

output :
1 2 3
*/

#include <iostream>
#include <algorithm>
#include <vector>
#include <queue>
using namespace std;

// 전역 변수 선언
int N, M;
int a, b;
vector<int> graph[32005];
int indeg[32005];
queue<int> que;


int main(){
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    freopen("input.txt", "r", stdin);

    cin >> N >> M;
    while (M--){
        cin >> a >> b;
        graph[a].push_back(b);
        indeg[b] += 1;
    }

    for (int i = 1; i <= N; i++){
        if (indeg[i] == 0){
            que.push(i);
        }
    }

    while (que.size() != 0){
        int cur = que.front(); que.pop();
        cout << cur << ' ';
        for (int nxt: graph[cur]){
            indeg[nxt] -= 1;
            if (indeg[nxt] == 0){
                que.push(nxt);
            }
        }
    }

    return 0;
}

File: 2021_summer_algorithm_study_intermediate/using_python/2252.py
"""
input :
3 2
1 3
2 3

output :
1 2 3
"""
import sys
from collections import deque
sys.stdin = open("input.txt", "r")
input = sys.stdin.readline

if __name__ == "__main__":
    N, M = map(int, input().rstrip().split())
    graph = [[] for _ in range(N + 1)]
    indeg = [0 for _ in range(N + 1)]
    for _ in range(M):
        a, b = map(int, input().rstrip().split())
        graph[a].append(b)
        indeg[b] += 1

    que = deque()
    for i in range(1, N + 1):
        if indeg[i] == 0:
            que.append(i)

    while len(que) != 0:
        cur = que.popleft()
        print(cur, end=' ')
        for nxt in graph[cur]:
            indeg[nxt] -= 1
            if indeg[nxt] == 0:
                que.append(nxt)
